simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/irr/ClockProPolicy.java

```diff
@@ -73,263 +70,556 @@
+  private int maxSize;
+  private int maxNonResSize;
 
-  // Maximum number of cold pages (adaptive):
+  private int sizeHot;
+  private int sizeResCold;
+  private int sizeNonResCold;
+  private int sizeInTest;
+  private int sizeFree;
+
+  // Target number of resident cold pages (adaptive):
   //  - increases when test page gets a hit
   //  - decreases when test page is removed
-  private int maximumColdSize;
+  private int coldTarget;
+  // {min,max}ResColdSize are boundary of coldTarget.
+  private int minResColdSize;
+  private int maxResColdSize;
 
-  private int sizeHot;
-  private int sizeCold;
-  private int sizeTest;
+  // Enable to print out the internal state
+  static final boolean debug = false;
 
   public ClockProPolicy(Config config) {
-    BasicSettings settings = new BasicSettings(config);
-    maximumSize = Ints.checkedCast(settings.maximumSize());
-    policyStats = new PolicyStats("irr.ClockPro");
-    data = new Long2ObjectOpenHashMap<>();
-    maximumColdSize = maximumSize;
-
-    // All the hands move in the clockwise direction
-    handHot = handCold = handTest = null;
+    ClockProSettings settings = new ClockProSettings(config);
+    this.maxSize = Ints.checkedCast(settings.maximumSize());
+    this.maxNonResSize = (int) (maxSize * settings.nonResidentMultiplier());
+    this.minResColdSize = (int) (maxSize * settings.percentMinCold());
+    if (minResColdSize < settings.lowerBoundCold()) {
+      minResColdSize = settings.lowerBoundCold();
+    }
+    this.maxResColdSize = (int) (maxSize * settings.percentMaxCold());
+    if (maxResColdSize > maxSize - minResColdSize) {
+      maxResColdSize = maxSize - minResColdSize;
+    }
+    this.policyStats = new PolicyStats("irr.ClockPro");
+    this.data = new Long2ObjectOpenHashMap<>();
+    this.coldTarget = minResColdSize;
+    this.listHead = this.handHot = this.handCold = this.handTest = null;
+    this.sizeFree = maxSize;
+    checkState(minResColdSize <= maxResColdSize);
   }
 
-  /** Returns all variations of this policy based on the configuration parameters. */
+  /**
+   * Returns all variations of this policy based on the configuration parameters.
+   */
   public static Set<Policy> policies(Config config) {
     return ImmutableSet.of(new ClockProPolicy(config));
   }
 
+  @Override
+  public PolicyStats stats() {
+    return policyStats;
+  }
+
+  @Override
+  public void finished() {
+    validateStatus();
+    validateClockStructure();
+    if (debug) {
+      printClock();
+    }
+  }
+
   @Override
   public void record(long key) {
+    policyStats.recordOperation();
     Node node = data.get(key);
-
     if (node == null) {
-      onMiss(key);
-    } else if (node.status == Status.TEST) {
-      onNonResidentHit(node);
-    } else {
+      node = new Node(key);
+      data.put(key, node);
+      onMiss(node);
+    } else if (node.isResident()) {
       onHit(node);
+    } else {
+      onMiss(node);
+    }
+    organizeHands();
```

Comment:
Does this violate the benefits of Clock for read concurrency?

The main benefit of Clock algorithms is to not modify central, shared state on reads by instead only updating per-entry metadata. This then leads to a higher cost on write when choose a victim. However this method updates the hands, e.g. it leads to `terminateTestPeriod` which modifies the linked lists. This can't be done safely or efficiently outside of a lock, so wouldn't it defeat the gains compared to using LIRS directly?

Reply:
Yes, it violates the benefits of clock, it's my mistake. I moved the organizeHands() into the onMiss() method. The test results are same because onHit() never change the status of clock entries so calling organizeHands() after onHit() not cause hands moving.